### ISSUE TITLE
Throwing errors with Apollo Server/Apollo Gateway

### ISSUE THREAD

**Issue by Tom-Pearce** [confuser/graphql-constraint-directive#131]
tldr; failed input validations not returning correct error code for Apollo Servers, thus preventing Apollo Gateway from displaying the error messages in a concise object.

Hello,

Currently implementing the constraint directive on an Apollo Server instance, that is acting as an auth microservice.

This then sits behind my Apollo-server-express gateway.

The issue I'm having is that when an input value doesn't meet with constraint criteria, the error messages arent bubbling up to the gateway, the way they should.

Looking at the [readme](https://github.com/confuser/graphql-constraint-directive#apollo-server-1), it says that for Apollo Server, it should be specifying the code [`BAD_USER_INPUT`](https://www.apollographql.com/docs/apollo-server/data/errors/#bad_user_input). However, when querying with invalid input, this is the response I get from my gateway:

```json
{
  "errors": [
    {
      "message": "500: Internal Server Error",
      "extensions": {
        "response": {
          "url": "http://localhost:4001/",
          "status": 500,
          "statusText": "Internal Server Error",
          "body": {
            "errors": [
              {
                "message": "Unexpected error value: Variable \"$input\" got invalid value \"thomas.pearceopenstudycollege.com\" at \"input.email\". Must be in email format",
                "extensions": {
                  "code": "INTERNAL_SERVER_ERROR",
                  "stacktrace": [
                    "GraphQLError: Unexpected error value: Variable \"$input\" got invalid value \"thomas.pearceopenstudycollege.com\" at \"input.email\". Must be in email format",
                    "    at ensureError (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/server/src/errorNormalize.ts:91:7)",
                    "    at ensureGraphQLError (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/server/src/errorNormalize.ts:98:24)",
                    "    at processGraphQLRequest (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/server/src/requestPipeline.ts:331:55)",
                    "    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
                    "    at async internalExecuteOperation (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/server/src/ApolloServer.ts:1216:12)",
                    "    at async runHttpQuery (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/server/src/runHttpQuery.ts:195:27)",
                    "    at async runPotentiallyBatchedHttpQuery (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/server/src/httpBatching.ts:75:12)",
                    "    at async ApolloServer.executeHTTPGraphQLRequest (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/server/src/ApolloServer.ts:995:14)"
                  ]
                }
              }
            ]
          }
        },
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "stacktrace": [
            "GraphQLError: 500: Internal Server Error",
            "    at RemoteGraphQLDataSource.errorFromResponse (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/datasources/RemoteGraphQLDataSource.ts:312:12)",
            "    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
            "    at async RemoteGraphQLDataSource.sendRequest (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/datasources/RemoteGraphQLDataSource.ts:201:15)",
            "    at async RemoteGraphQLDataSource.process (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/datasources/RemoteGraphQLDataSource.ts:157:22)",
            "    at async sendOperation (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/executeQueryPlan.ts:417:22)",
            "    at async /Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/executeQueryPlan.ts:301:41",
            "    at async executeNode (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/executeQueryPlan.ts:239:9)",
            "    at async /Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/executeQueryPlan.ts:80:27",
            "    at async /Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/index.ts:823:28",
            "    at async execute (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/apollo-server-core/src/requestPipeline.ts:499:14)",
            "    at async processGraphQLRequest (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/apollo-server-core/src/requestPipeline.ts:407:22)",
            "    at async processHTTPRequest (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/apollo-server-core/src/runHttpQuery.ts:434:24)"
          ]
        }
      }
    }
  ],
  "data": {
    "createUser": null
  }
}
```

According to this, the `INTERNAL_SERVER_ERROR` code is being thrown instead, which is what I think is preventing the actual error messages from propagating up to the gateway response.

Instead I would have to do something like the following to access them: `response.errors.extensions.response.body.errors.message`.

In one of my resolvers (on the same microservice) throws an error like so:
```typescript
throw new GraphQLError('An account already exists for the specified email.', {
    extensions: {
        code: 'BAD_USER_INPUT'
    }
});
```
The error bubbles up correctly:
```json
{
  "errors": [
    {
      "message": "An account already exists for the specified email.",
      "extensions": {
        "code": "BAD_USER_INPUT",
        "stacktrace": [
          "GraphQLError: An account already exists for the specified email.",
          "    at createUser (/Users/tompearce/Documents/OSC/new-stack/osc/packages/osc-api-auth/src/resolvers/resolver.ts:88:23)"
        ],
        "serviceName": "OSC-Auth",
        "exception": {
          "message": "An account already exists for the specified email.",
          "locations": [
            {
              "line": 1,
              "column": 59
            }
          ],
          "path": [
            "createUser"
          ],
          "stacktrace": [
            "GraphQLError: An account already exists for the specified email.",
            "    at downstreamServiceError (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/executeQueryPlan.ts:622:12)",
            "    at /Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/executeQueryPlan.ts:432:9",
            "    at Array.map (<anonymous>)",
            "    at sendOperation (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/executeQueryPlan.ts:431:38)",
            "    at processTicksAndRejections (node:internal/process/task_queues:95:5)",
            "    at async /Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/executeQueryPlan.ts:301:41",
            "    at async executeNode (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/executeQueryPlan.ts:239:9)",
            "    at async /Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/executeQueryPlan.ts:80:27",
            "    at async /Users/tompearce/Documents/OSC/new-stack/osc/node_modules/@apollo/gateway/src/index.ts:823:28",
            "    at async execute (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/apollo-server-core/src/requestPipeline.ts:499:14)",
            "    at async processGraphQLRequest (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/apollo-server-core/src/requestPipeline.ts:407:22)",
            "    at async processHTTPRequest (/Users/tompearce/Documents/OSC/new-stack/osc/node_modules/apollo-server-core/src/runHttpQuery.ts:434:24)"
          ]
        }
      }
    }
  ],
  "data": {
    "createUser": null
  }
}
```

I had a quick look through the source code, but couldn't find anything referencing the `BAD_USER_INPUT` code.

**Comment by confuser** [comment]
Given GraphQL now supports this (it didn't before!) we should probably remove the custom ValidationError object and always throw a GraphQLError

**Comment by velias** [comment]
@confuser Agree, we should use GraphQL Error. I can make this change
